src/building/bootstrapping.md

```diff
@@ -108,50 +110,92 @@ contribution [here][bootstrap-build].
 
 ## Understanding stages of bootstrap
 
-This is a detailed look into the separate bootstrap stages. When running
-`x.py` you will see output such as:
-
-```txt
-Building stage0 std artifacts
-Copying stage0 std from stage0
-Building stage0 compiler artifacts
-Copying stage0 rustc from stage0
-Building LLVM for x86_64-apple-darwin
-Building stage0 codegen artifacts
-Assembling stage1 compiler
-Building stage1 std artifacts
-Copying stage1 std from stage1
-Building stage1 compiler artifacts
-Copying stage1 rustc from stage1
-Building stage1 codegen artifacts
-Assembling stage2 compiler
-Uplifting stage1 std
-Copying stage2 std from stage1
-Generating unstable book md files
-Building stage0 tool unstable-book-gen
-Building stage0 tool rustbook
-Documenting standalone
-Building rustdoc for stage2
-Documenting book redirect pages
-Documenting stage2 std
-Building rustdoc for stage1
-Documenting stage2 whitelisted compiler
-Documenting stage2 compiler
-Documenting stage2 rustdoc
-Documenting error index
-Uplifting stage1 rustc
-Copying stage2 rustc from stage1
-Building stage2 tool error_index_generator
-```
+### Overview
+
+This is a detailed look into the separate bootstrap stages.
+
+The convention `x.py` uses is that:
+- A `--stage N` flag means to run the stage N compiler (`stageN/rustc`).
+- A "stage N artifact" is an artifact that is _produced_ by the stage N compiler.
+- The "stage (N+1) compiler" is assembled from "stage N artifacts". This
+  process is called _uplifting_.
+
+For example, `x.py build --stage 0` means to build with the beta `rustc`, and
+`test --stage 0` isn't really meaningful (it's not running tests on your changes,
+but on `beta`, so it always fails). Similarly, `doc --stage 0` means to
+document using the beta `rustdoc`.
+
+Note that this implies the stage N compiler is _not_ the same as the compiler
+built by `build --stage N compiler/rustc` -- that's 'stage N artifacts'
+('the compiler built by stage N').
+
+In short, _stage 0 uses the stage0 compiler to create stage0 artifacts which
+will later be uplifted to stage1_.
+
+In each stage, two major steps are performed:
+
+1. `std` is compiled by the stage N compiler.
+2. That `std` is linked to programs built by the stage N compiler, including
+   the stage (N+1) compiler.
+
+This is somewhat intuitive if one thinks of the stage (N+1) compiler as "just"
+another program we are building with the stage N compiler:
+`build --stage N compiler/rustc` is linking the stage (N+1) compiler to the `std`
+built by the stage N compiler.
```

Comment:
Given the off-by-1 here everywhere, wouldn't it be better to always say "stage N artifacts" instead of "stage N+1 compiler"? That also aligns better with "stage N std" below.